### ISSUE TITLE
False positve CS0165 'Use of unassigned local variable' on an value type like int, bool, ...

### ISSUE THREAD

**Issue by mikeKuester** [dotnet/roslyn#69018]
As the C# docs mention - value types have always an default value:

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/integral-numeric-types

> The default value of each integral type is zero, 0.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/default-values

> Any built-in integral numeric type | 0 (zero)
> Any built-in floating-point numeric type | 0 (zero)
> bool | false
> char | '\0' (U+0000)

So although these types have a default value and are therefore always initialized, there is the compiler error "CS0165 Use of unassigned local variable" if I create a local integer variable without explicitly assigning a value. Instead of that, I can create a field in a class without that compiler error.

**Version Used**: 
- Microsoft (R) Visual C# Compiler Version 4.6.0-3.23259.8 (c3cc1d0c)
- Microsoft Visual Studio Enterprise 2022 Version 17.6.5

**Steps to Reproduce**:

Create a local variable of type int, bool, double, ... and use it without initializing it with its default value.
```csharp
int i;
int j;

// Error CS0165 Use of unassigned local variable 'i'
j = i;
```

**Expected Behavior**:

```csharp
int i;
int j;

// no error
j = i;
```

**Actual Behavior**:

```csharp
int i;
int j;

// Error CS0165 Use of unassigned local variable 'i'
j = i;
```

But an uninitialized field is valid:

```csharp
internal class IntProperty
{
	// no error
	int myVar;

	public int MyProperty
	{
		get { return myVar; }
		set { myVar = value; }
	}
}
```

Either a value type has a default value, as documented everywhere, then it does not have to be initialized explicitly (see "CA1805: Do not initialize unnecessarily"). A different treatment between a local variable or a field variable of a class is not understandable.

If a value type has no default value, then it must be initialized everywhere and the documentation must be adapted accordingly.

**Comment by CyrusNajmabadi** [comment]
The language and docs are both correct. The docs are explaining what the default value of a value type is. I.e. what is value would be if you wrote:

`V v = default`.  

If V is a reference type, the default value is null. If it is a numeric value type it is zero. Etc. Etc.

That the default value is zero does not mean it has that value when *not* initialized.  A variable may or may not be initialized depending on several rules. 

For example, a field variable is auto initialized.  An empty struct is considered initialized.  Variables *within* an array are initialized. Etc. Etc.

However, locals are not automatically initialized. They must be done so manually.

**Comment by CyrusNajmabadi** [comment]
> But an uninitialized field is valid:

Correct. Compiler initializes these on your behalf. It's equivalent to adding an `= default;` to each field. 

> A different treatment between a local variable or a field variable of a class is not understandable.

You're welcome to think that. But that is quite intentionally how the language is defined. If you want the language to change, the requests for that are made as discussions at dotnet/csharplang.  As of now though, the compiler is correctly implementing the intentionally specified rules of the language.

**Comment by mikeKuester** [comment]
@CyrusNajmabadi Thanks for the explanation. One question:

If I debug through this code and step to this breakpoint ...

```csharp
int i;
int j;

i = 1; <== breakpoint

Console.WriteLine($"Hello: {i}, {j}");
``` 
both local variables are inirialized with 0. Why is that? You wrote that they will be uninitialized?

**Comment by CyrusNajmabadi** [comment]
Can you link me to a repro file?

Note: there are multiple concepts of 'initialized ' here. The first is whether or not the *language* considers something initialized.  The second is how the runtime chooses to work. The runtime is free, for example, to take those variables and give them initial default values (or not). If you're on a debugger, you are seeing what the runtime has chosen, which isn't itself relevant to the language here.

**Comment by CyrusNajmabadi** [comment]
When I try your example, I get a proper error from the compiler:

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAlgHYAu6zA3I00W04ArXkn7Fm6ALzo43dAHoF6ADxSZAIwBOAUwCGAawAOAe0F9+UOAE5qAEgBEACR0AbVyZD5mAXwA0+EI+DrSiTD7IPkA===

error CS0165: Use of unassigned local variable 'j'

**Comment by mikeKuester** [comment]
Ok, thanks for your time. I think I mixed the language and the runtime here. 🤔

**Comment by CyrusNajmabadi** [comment]
You're welcome!
